File: swi-prolog/n_queens.pl
/*
  Solve the N Quenns problem. From the Swi-Prolog documentation:
     https://www.swi-prolog.org/pldoc/man?section=clpfd-n-queens
*/

n_queens(N, Qs) :-
        length(Qs, N),
        Qs ins 1..N,
        safe_queens(Qs).

safe_queens([]).
safe_queens([Q|Qs]) :- safe_queens(Qs, Q, 1), safe_queens(Qs).

safe_queens([], _, _).
safe_queens([Q|Qs], Q0, D0) :-
        Q0 #\= Q,
        abs(Q0 - Q) #\= D0,
        D1 #= D0 + 1,
    safe_queens(Qs, Q0, D1).

File: swi-prolog/n_queens.py
"Example calling Swi-Prolog N-Queens program from Python"

from swiplserver import PrologMQI
from pprint import pprint

with PrologMQI() as mqi:
  with mqi.create_thread() as prolog_thread:
    prolog_thread.query("use_module(library(clpfd)).")
    prolog_thread.query("[n_queens].")
    result = prolog_thread.query("n_queens(8, Qs), label(Qs).")
    pprint(result)
    print(len(result))

